### ISSUE TITLE
git_merge_abort() doesn't do the reset

### ISSUE THREAD

**Issue by jennybc** [r-lib/gert#66]
`gert::git_merge_abort()` doesn't do what `git merge --abort` does.

It feels like it just changes some internal repo state (?) but doesn't finish the job by doing `git reset --hard HEAD`. I.e. the conflicted files still have all their conflict markers after `git_merge_abort()`.

**Comment by jennybc** [comment]
Reprex (using usethis):

In an R session:

```
library(usethis)
create_project("~/tmp/susan")
```

Then, in that new project named "susan":

```
library(usethis)
use_readme_md()
use_git()
use_git_ignore("reprex*")
use_github(private = TRUE)
```

Locally, still in "susan":
``` r
library(gert)
#> Linking to libgit2 v0.28.4, ssh support: YES
#> Global config: /Users/jenny/resources/system-setup/dotfiles/DOTgitconfig
#> Default user: Jenny Bryan <jenny.f.bryan@gmail.com>

git_status()
#> # A tibble: 0 x 3
#> # … with 3 variables: file <chr>, status <chr>, staged <lgl>

cat("apple", sep = "\n", file = "README.md", append = TRUE)
git_status()
#> # A tibble: 1 x 3
#>   file      status   staged
#>   <chr>     <chr>    <lgl> 
#> 1 README.md modified FALSE

git_add("README.md")
#> # A tibble: 1 x 3
#>   file      status   staged
#>   <chr>     <chr>    <lgl> 
#> 1 README.md modified TRUE
git_commit("apple")
#> [1] "3101440004dcf7fb471ddb9004b14fdd0a520a01"

# Meanwhile, on GitHub, edit README.md in the browser
# Append the line "banana"

git_fetch()
git_merge("origin/master")
#> Merge has resulted in merge conflict(s).
git_status()
#> # A tibble: 1 x 3
#>   file      status     staged
#>   <chr>     <chr>      <lgl> 
#> 1 README.md conflicted FALSE

readLines("README.md")
#>  [1] ""                                                
#>  [2] "# susan"                                         
#>  [3] ""                                                
#>  [4] "<!-- badges: start -->"                          
#>  [5] "<!-- badges: end -->"                            
#>  [6] ""                                                
#>  [7] "The goal of susan is to ..."                     
#>  [8] ""                                                
#>  [9] "<<<<<<< HEAD"                                    
#> [10] "apple"                                           
#> [11] "======="                                         
#> [12] "banana"                                          
#> [13] ">>>>>>> 18ff64d020952d2271b878626e47fbabeaf3bb2a"

# do `git merge --abort` in the shell
system("git merge --abort")
git_status()
#> # A tibble: 0 x 3
#> # … with 3 variables: file <chr>, status <chr>, staged <lgl>
readLines("README.md")
#> [1] ""                            "# susan"                    
#> [3] ""                            "<!-- badges: start -->"     
#> [5] "<!-- badges: end -->"        ""                           
#> [7] "The goal of susan is to ..." ""                           
#> [9] "apple"

# do the merge again but use git_merge_abort()
git_merge("origin/master")
#> Merge has resulted in merge conflict(s).

git_status()
#> # A tibble: 1 x 3
#>   file      status     staged
#>   <chr>     <chr>      <lgl> 
#> 1 README.md conflicted FALSE
readLines("README.md")
#>  [1] ""                                                
#>  [2] "# susan"                                         
#>  [3] ""                                                
#>  [4] "<!-- badges: start -->"                          
#>  [5] "<!-- badges: end -->"                            
#>  [6] ""                                                
#>  [7] "The goal of susan is to ..."                     
#>  [8] ""                                                
#>  [9] "<<<<<<< HEAD"                                    
#> [10] "apple"                                           
#> [11] "======="                                         
#> [12] "banana"                                          
#> [13] ">>>>>>> 18ff64d020952d2271b878626e47fbabeaf3bb2a"

git_merge_abort()
#> NULL

git_status()                # still shows README.me is conflicted
#> # A tibble: 1 x 3
#>   file      status     staged
#>   <chr>     <chr>      <lgl> 
#> 1 README.md conflicted FALSE
readLines("README.md")      # conflict markers still present
#>  [1] ""                                                
#>  [2] "# susan"                                         
#>  [3] ""                                                
#>  [4] "<!-- badges: start -->"                          
#>  [5] "<!-- badges: end -->"                            
#>  [6] ""                                                
#>  [7] "The goal of susan is to ..."                     
#>  [8] ""                                                
#>  [9] "<<<<<<< HEAD"                                    
#> [10] "apple"                                           
#> [11] "======="                                         
#> [12] "banana"                                          
#> [13] ">>>>>>> 18ff64d020952d2271b878626e47fbabeaf3bb2a"

system("git reset --hard HEAD")

git_status()                # clean
#> # A tibble: 0 x 3
#> # … with 3 variables: file <chr>, status <chr>, staged <lgl>
readLines("README.md")      # back to pre-merge state
#> [1] ""                            "# susan"                    
#> [3] ""                            "<!-- badges: start -->"     
#> [5] "<!-- badges: end -->"        ""                           
#> [7] "The goal of susan is to ..." ""                           
#> [9] "apple"
```

<sup>Created on 2020-06-27 by the [reprex package](https://reprex.tidyverse.org) (v0.3.0.9001)</sup>

**Comment by jeroen** [comment]
Thanks! I didn't know merge-abort would also hard-reset. Could you test if it is fixed now?

**Comment by jennybc** [comment]
Something seems wonky. I see this:

```
> git_status()
# A tibble: 1 x 3
  file      status     staged
  <chr>     <chr>      <lgl> 
1 README.md conflicted FALSE 
> git_merge_abort()
Error in git_open(repo) : 
  promise already under evaluation: recursive default argument reference or earlier problems?
Run `rlang::last_error()` to see where the error occurred.
> le
<error/rlang_error>
Error in git_open(repo) : 
  promise already under evaluation: recursive default argument reference or earlier problems?
Backtrace:
 1. gert::git_merge_abort()
 2. gert::git_reset_hard(repo = repo) /Users/jenny/rrr/gert/R/merge.R:93:2
 3. gert:::git_reset("hard", ref = ref, repo = repo) /Users/jenny/rrr/gert/R/rebase.R:57:2
 4. gert::git_open(repo) /Users/jenny/rrr/gert/R/rebase.R:75:2
Run `rlang::last_trace()` to see the full context.
```

I can't debug further myself right now as I don't strictly need this for my rough demo tomorrow.

**Comment by jennybc** [comment]
As for the mechanics of this, I don't know if `git reset --hard HEAD` is the official solution (but it sure seems to have the right outcome).

I don't know if libgit2 docs or examples offers any help on this? Here's one morsel of info I came across while researching something else:

> `ORIG_HEAD` is created by commands that move your `HEAD` in a drastic way, to record the position of the HEAD before their operation, so that you can easily change the tip of the branch back to the state before you ran them.

Found here:

https://git-scm.com/docs/gitrevisions#Documentation/gitrevisions.txt-emltrefnamegtemegemmasterememheadsmasterememrefsheadsmasterem

(I have no idea what's going on with that link! But it works.)

I don't know if `ORIG_HEAD` is lying around for you, but that's another candidate for the target of the hard reset. I only raise this because of some vague concern that `HEAD` might not always be the right target and `ORIG_HEAD` would be safer? But `HEAD` sure seems like what we want. If people merged with uncommitted changes, it might not be possible to restore everything, but that's basically a known risk in that scenario.
